ACPs/113-provable-randomness/README.md

```diff
@@ -0,0 +1,157 @@
+
+A solution for verifiable random number generation within Avalanche solves these problems. It provides fair randomness functionality across the chains, at no additional cost. This paves the way for a more efficient Avalanche ecosystem.
+
+## Specification
+
+### Changes Summary
+
+Existing avalanche protocol breaks the block building into two parts : external and internal. The external block is the SnowMan++ block, whereas the internal block is the actual virtual machine block.
+
+To support randomness, a BLS based VRF implementation is used, that would be recursively signing its own signatures as its message. Since the BLS signatures are deterministic, they provide a great way to construct a reliable VRF.
+
+For proposers that do not have a BLS key associated with their node, the hash of the signature from the previous round.
+
+In order to bootstrap the signatures chain, a missing signature would be replaced with a byte slice that is the hash product of a verifiable and trustable seed.
+
+The changes proposed here would affect the way a block is being validated. Therefore, when this change gets implemented, it needs to be deployed as a mandatory upgrade.
+
+```
+		+-----------------------+           +-----------------------+
+		|  Block n              | <-------- |  Block n+1            |
+		+-----------------------+           +-----------------------+
+		|  VRF-Sig(n)           |           |  VRF-Sig(n+1)         |
+		|  ...                  |           |  ...                  |
+		+-----------------------+           +-----------------------+
+
+		+-----------------------+           +-----------------------+
+		|  VM n                 |           |  VM n+1               |
+		+-----------------------+           +-----------------------+
+		|  VRF-Out(n)           |           |  VRF-Out(n+1)         |
+		+-----------------------+           +-----------------------+
+
+		VRF-Sig(n+1) = Sign(VRF-Sig(n), Block n+1 proposer's BLS key)
+		VRF-Out(n) = Hash(VRF-Sig(n))
+```
+
+### Changes Details
+
+#### Step 1. Adding BLS signature to proposed blocks
+
+```golang
+type statelessUnsignedBlock struct {
+	…
+	vrfSig    []byte `serialize:”true”`
+}
+```
+
+#### Step 2. Populate signature
+
+When a block proposer attempts to build a new block, it would need to use the parent block as a reference.
+
+The `vrfSig` field within each block is going to be daisy-chained to the `vrfSig` field from it's parent block.
+
+Populating the `vrfSig` would following this logic:
+
+1. The current proposer has a BLS key
+   
+	a. If the parent block has an empty `vrfSig` signature, that signature would be signed using the proposer’s BLS key. This is the base case.
+
+	b. If the parent block does not have an empty `vrfSig` signature, the proposer would sign the bootStrappingBlockSignature with its BLS key. See the bootStrappingBlockSignature details below.
```

Comment:
Aren't these cases swapped? You bootstrap if the `vrfSig` *is* empty right? If its not then you just chain sign.

Reply:
correct. fixed.